### PR TITLE
Initial operator with controller creating cluster tasks

### PR DESCRIPTION
**Pull request by JaydipGabani**

### CODE REVIEW

File: config/manifests/bases/mtk-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    alm-examples: '[]'
+    capabilities: Basic Install
+  name: mtk-operator.v0.0.0
```
Comment (djzager) [inline config/manifests/bases/mtk-operator.clusterserviceversion.yaml:7]: I really like what mig-operator did using `v99.0.0` for development. We should do the same.

References:
* https://github.com/konveyor/mig-operator/blob/c8f97a444bc32073a4b7e6d8f8f02df5a6de47c0/deploy/olm-catalog/bundle/manifests/crane-operator.v99.0.0.clusterserviceversion.yaml#L5 
* https://github.com/openshift/assisted-service/blob/364b9d71b11250ae536787ef354c2d7cb516bb7c/config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml#L18

File: config/rbac/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+- role_binding.yaml
+- leader_election_role.yaml
+- leader_election_role_binding.yaml
+# Comment the following 4 lines if you want to disable
+# the auth proxy (https://github.com/brancz/kube-rbac-proxy)
+# which protects your /metrics endpoint.
+- auth_proxy_service.yaml
+- auth_proxy_role.yaml
+- auth_proxy_role_binding.yaml
+- auth_proxy_client_clusterrole.yaml
```
Comment (djzager) [inline config/rbac/kustomization.yaml:18]: I **_believe_** that we can comment these out.

File: controllers/operatorconfig_controller.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
+func (r *OperatorConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// your logic here
+
```
Comment (djzager) [inline controllers/operatorconfig_controller.go:63]:
```suggestion
```

File: controllers/operatorconfig_controller.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import (
+	"context"
+	"github.com/ghodss/yaml"
+	"github.com/go-logr/logr"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	config2 "sigs.k8s.io/controller-runtime/pkg/client/config"
```
Comment (djzager) [inline controllers/operatorconfig_controller.go:28]:
```suggestion
	"sigs.k8s.io/controller-runtime/pkg/client/config"
```

File: controllers/operatorconfig_controller.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	errs := []error{}
+
+	var data []byte
+	data, err = ioutil.ReadFile("clustertasks.yaml")
+	if err != nil {
+		return err
+	}
+
+	config := config2.GetConfigOrDie()
+	clientset, err := versioned.NewForConfig(config)
```
Comment (djzager) [inline controllers/operatorconfig_controller.go:87]:
```suggestion
	clientset, err := versioned.NewForConfig(config.GetConfigOrDie())
```

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+vet: ## Run go vet against code.
+	go vet ./...
+
+.PHONY: test
+test: manifests generate fmt vet envtest ## Run tests.
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test ./... -coverprofile cover.out
+
+.PHONY: clustertasks
+clustertasks:
+	mkdir -p deploy/artifacts/ && touch deploy/artifacts/clustertasks.yaml | $(KUSTOMIZE) build $(CLUSTERTASK_SOURCE) > deploy/artifacts/clustertasks.yaml
```
Comment (djzager) [inline Makefile:104]: Nice. I think I would just call this a `manifests.yaml`.

File: controllers/operatorconfig_controller.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		}
+		obj, err := clientset.TektonV1beta1().ClusterTasks().Create(ctx, &clustertasks, metav1.CreateOptions{})
+		if err != nil {
+			errs = append(errs, err)
+		}
+		log.Info("Cluster task created", obj.Name, obj.Namespace)
+	}
+	if len(errs) != 0 {
+		return errors.NewAggregate(errs)
+	}
```
Comment (djzager) [inline controllers/operatorconfig_controller.go:105]: I believe this works for the initial implementation. We will want to look into handling the scenario where we have more than `ClusterTasks` in the manifest.yaml

**Review by djzager — CHANGES_REQUESTED** [review]
Looking good.

File: controllers/operatorconfig_controller.go (hunk 7 above)
Comment (JaydipGabani) [reply controllers/operatorconfig_controller.go:105]: +1

Comment (JaydipGabani) [reply controllers/operatorconfig_controller.go:102]: I initially intended the solution to be "kind" independent, but its a little difficult to implement and figure out "kind" on the runtime. Still working through it to figure out what would be the best approach to solve this. Welcoming any ideas to implement this behavior or another approach.

**Review by djzager — APPROVED** [review]

**Review by sseago — APPROVED** [review]

File: bundle/metadata/dependencies.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+dependencies:
+  - type: olm.package
+    value:
+      packageName: openshift-pipelines-operator-rh
+      version: "=1.2.3"
```
Comment (djzager) [inline bundle/metadata/dependencies.yaml:5]:
```suggestion
      version: ">=1.2.3"
```

**Review by djzager — CHANGES_REQUESTED** [review]

**Pull request merged**